pkg/handlers/v1/storage/db.go

```diff
@@ -0,0 +1,309 @@
+	LogFn domain.LogFn // also a seam
+
+	once sync.Once
+}
+
+// Init initializes a connection to a Postgres database according to the environment variables POSTGRES_HOST, POSTGRES_PORT, POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DATABASE
+func (db *DB) Init(ctx context.Context, postgresConfig *domain.PostgresConfig) error {
+	var initerr error
+	db.once.Do(func() {
+
+		host := postgresConfig.Hostname
+		port := postgresConfig.Port
+		user := postgresConfig.Username
+		password := postgresConfig.Password
+		dbname := postgresConfig.DatabaseName
+
+		logger := db.LogFn(ctx)
+
+		if db.sqldb == nil {
+			sslmode := "disable"
+			if host != "localhost" {
+				sslmode = "enable"
+			}
+			psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+				"password=%s dbname=%s sslmode=%s",
+				host, port, user, password, dbname, sslmode)
+			pgdb, err := sql.Open("postgres", psqlInfo)
+			if err != nil {
+				logger.Error(logs.DBOpenError{Reason: err.Error()})
+				initerr = err
+			}
+
+			err = pgdb.Ping()
+			if err != nil {
+				logger.Error(logs.DBPingError{Reason: err.Error()})
+				initerr = err
+			}
+
+			db.sqldb = pgdb
+		}
+	})
+	return initerr
+}
+
+// StoreCloudAsset an implementation of the Storage interface that records to a database
+func (db *DB) StoreCloudAsset(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		switch err {
+		case nil:
+			err = tx.Commit()
+		default:
+			err = tx.Rollback()
+		}
+	}()
+
+	if err = db.saveResource(ctx, cloudAssetChanges); err != nil {
+		return err
+	}
+
+	for _, val := range cloudAssetChanges.Changes {
+		err = db.recordNetworkChanges(ctx, cloudAssetChanges.ARN, cloudAssetChanges.ChangeTime, val, tx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (db *DB) saveResource(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	// You won't get an ID back if nothing was done.  Also, this lib won't return the ID anyway even without the "ON CONFLICT DO NOTHING".
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	sqlStatement := fmt.Sprintf(`INSERT INTO %s VALUES ($1, $2, $3, $4, $5) ON CONFLICT DO NOTHING RETURNING id`, tableAWSResources)  // nolint
+
+	if _, err := db.sqldb.ExecContext(ctx, sqlStatement, cloudAssetChanges.ARN, cloudAssetChanges.AccountID, cloudAssetChanges.Region, cloudAssetChanges.ResourceType, cloudAssetChanges.Tags); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (db *DB) recordNetworkChanges(ctx context.Context, resourceID string, timestamp time.Time, changes domain.NetworkChanges, tx *sql.Tx) error {
+
+	for _, hostname := range changes.Hostnames {
+		err := db.insertHostname(ctx, hostname, tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, val := range changes.PrivateIPAddresses {
+		isJoin := false
+		if changes.ChangeType == "ADDED" {
```

Comment:
For all cases of string comparison, I'd recommend using a constant value for the sentinel and using https://golang.org/pkg/strings/?m=all#EqualFold rather than equality to ensure that we account for input that isn't cased the way we expect.